src/py21cmfast/src/HaloBox.c

```diff
@@ -699,12 +699,19 @@ int ComputeHaloBox(double redshift, UserParams *user_params, CosmoParams *cosmo_
         //Since we need the average turnover masses before we can calculate the global means, we do the CMF integrals first
         //Then we calculate the expected UMF integrals before doing the adjustment
         if(flag_options->FIXED_HALO_GRIDS){
+            M_max = M_MAX_INTEGRAL;
             set_fixed_grids(M_min, M_max, ini_boxes, perturbed_field, previous_spin_temp, previous_ionize_box, &hbox_consts, grids, &averages_box, true);
         }
         else{
             //set below-resolution properties
             if(user_params->AVG_BELOW_SAMPLER && M_min < user_params->SAMPLER_MIN_MASS){
-                set_fixed_grids(M_min, user_params->SAMPLER_MIN_MASS, ini_boxes,
+                if (flag_options->HALO_STOCHASTICITY){
+                    M_max = user_params->SAMPLER_MIN_MASS;
+                }
+                else {
+                    M_max = RtoM(L_FACTOR*user_params->BOX_LEN/user_params->DIM);
```

Comment:
There's one edge case that we're still missing here I think, which is when the cell size is so ridiculously small that it goes below the minimum source mass. in which case we can just skip the integrated component entirely. While it *does* get caught in the integrals themselves we probably don't want to go through the whole setup.